text/2019-11-07-raft-witness.md

```diff
@@ -0,0 +1,85 @@
+# Raft Witness
+
+## Summary
+
+This RFC proposes a new role Witness to Raft, and describes how it could benefit us.
+
+## Motivation
+
+To tolerate 2 replicas fail, Raft needs 5 replicas, which requires a huge energy
+footprint. However, with witness Raft could only need 4 or even 3 replicas to
+tolerate 2 fail, and the price is latency becomes higher if some peers are
+isolated, or need more time to recover from 2 peers fail. Considering they are not
+common cases, we can introduce Witness into Raft.
+
+## Dtailed Design
+
+### The Original Article
+
+Witness is firstly described in this [article][reduce-the-energy-footprint], and is
+already implemented in [dragonboat](https://github.com/lni/dragonboat). According to
+the article, Witness is like normal Voter in Raft, except
+
+* Witness can't start new elections;
+* Leader only replicates metadata (term, index) to Witness without real entries.
+
+So Witness can reduce network traffic and storage usage in a Raft group.
+
+However, there are some issues about about the design. For example (described in the article),
+suppose there is a Raft group with Voters `A, B, C, D` and Witness `E`. `A, B, E` can
+hold a quorum to commit some entries. But if `A, B` fail, `C, D, E` can't elect a new
+leader forever because `E` will never start new elections, and it will always reject
```

Comment:
> ...`E` will never start new elections

I don't think the improvement below can solve the issue.

Reply:
Why not? If `A` and `B` fails, `C` or `D` will start a new election, and the witness `E` won't reject their vote messages. So `C` or `D` must can become new leader.